### ISSUE TITLE
Adopted documents: support "stitch" command for "alongside wrap" of two languages

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/metanorma-standoc#742]
This issue is to support "alongside wrap" detailed at https://github.com/metanorma/metanorma-bsi/issues/2#issuecomment-852269964

Alongside wrap. A wraps B. B is unmodified. A provides additional content C where C corresponds to a transformation of B (e.g. a translation of B). A provides additional content outside of B and C.

- EVS EN ISO (alongside wrap). It provides the Estonian translation of the ISO content, ISO Foreword, etc and also translation of the EN Foreword, etc.

The aligning/stitching command is called `stitch` below.
This is all speculative so the actual commands can differ in real usage.

### Alongside wrap

en-iso-44001-english.adoc
```adoc
= EN ISO 44001
:lang: english

.EN Foreword
...

```

en-iso-44001-estonian.adoc
```adoc
= EN ISO 44001
:lang: estonian

.EN Foreword in Estonian
...

```

evs-en-iso-44001.adoc
```adoc
= EVS EN ISO 44001

.EVS Foreword
...

stitch::[en-iso-44001-english.adoc,en-iso-44001-estonian.adoc]

```

## Images

Cover:
<img width="562" alt="Screenshot 2023-02-14 at 12 47 03 AM" src="https://user-images.githubusercontent.com/11865/218519439-f91a0638-2d90-4de1-86a1-8202bd4c942c.png">

National foreword with side by side translation:
<img width="560" alt="Screenshot 2023-02-14 at 12 47 13 AM" src="https://user-images.githubusercontent.com/11865/218519489-2882a76f-e728-42b0-a069-d5c5c6178e3e.png">

European cover:
<img width="555" alt="Screenshot 2023-02-14 at 12 47 31 AM" src="https://user-images.githubusercontent.com/11865/218519566-c27c2483-abe1-430c-9b71-e0a5efc5f830.png">

TOC side by side:
<img width="550" alt="Screenshot 2023-02-14 at 12 47 46 AM" src="https://user-images.githubusercontent.com/11865/218519634-12015f8b-a39e-4775-bcfb-8c3f6da7f014.png">

European foreword side by side:
<img width="548" alt="Screenshot 2023-02-14 at 12 48 39 AM" src="https://user-images.githubusercontent.com/11865/218519846-a72b4246-f6eb-4531-bfb2-93320c2713c0.png">

Content side by side:
<img width="547" alt="Screenshot 2023-02-14 at 12 48 56 AM" src="https://user-images.githubusercontent.com/11865/218519924-99337cd9-e579-457e-a8b4-843b2adc8379.png">

Annex with Estonian table first:
<img width="545" alt="Screenshot 2023-02-14 at 12 49 22 AM" src="https://user-images.githubusercontent.com/11865/218520040-e66489ac-b7c9-4cc3-b5bd-3b6a844c7cf7.png">

Table continue in Estonian:
<img width="539" alt="Screenshot 2023-02-14 at 12 50 08 AM" src="https://user-images.githubusercontent.com/11865/218520225-5e2e1367-7a62-4b78-bf44-6ba40f83b938.png">

English table:
<img width="542" alt="Screenshot 2023-02-14 at 12 50 25 AM" src="https://user-images.githubusercontent.com/11865/218520323-4ccc6f65-1c22-4da5-b79e-3f388b0e313a.png">

Bibliography with heading in 2 languages, content only English:

<img width="548" alt="Screenshot 2023-02-14 at 12 51 07 AM" src="https://user-images.githubusercontent.com/11865/218520470-70372c80-1022-4f7a-ab28-7a9cf2ddc84d.png">

Index in Estonian:

<img width="547" alt="Screenshot 2023-02-14 at 12 51 54 AM" src="https://user-images.githubusercontent.com/11865/218520635-48988fee-4e6c-46cf-b216-6f3cec97df2c.png">

Index in English:
<img width="539" alt="Screenshot 2023-02-14 at 12 52 02 AM" src="https://user-images.githubusercontent.com/11865/218520666-9ded2042-35de-410e-9f67-818f0bf81b39.png">

Back cover:
<img width="536" alt="Screenshot 2023-02-14 at 12 52 14 AM" src="https://user-images.githubusercontent.com/11865/218520703-6c484f9c-0aa0-450b-9ef4-8e67a5534f63.png">

_Originally posted by @ronaldtse in https://github.com/metanorma/metanorma-bsi/issues/2#issuecomment-852269964_

**Comment by opoudjis** [comment]
So, stitch:[] is a minimally thought out request to embed two documents simultaneously, where the second document is a translation of the first, and the alignment between the two is to be realised by magic.

It will not be realised by magic. Of the features specified in https://github.com/metanorma/metanorma-standoc/issues/420, the multilingual-rendering attributes will still need to be inserted into the two stitched documents, as will :align-cross-elements: : these documents will be marked up for bilingual alignment. At most, the stitching will assume that the clause structure of the two documents is identical, and where it is, it will insert tag in the two corresponding clauses to line them up explicitly.

**Comment by ronaldtse** [comment]
"Magic" is defined here for creating a element correspondence.

The element correspondence between two languages is either manually encoded (e.g. anchors "id1@en" matches "id2@jp") or automatically matched according to sequence.

There are cases of exceptions such as:
* 1 Document-A element corresponds to multiple, sequential, Document-B elements, vice versa
* 1 Document-A element corresponds to no element in Document-B (i.e. it is not translated into Document-B), vice versa

**Comment by opoudjis** [comment]
I will get some bioinformatics algorithm or other to do best-case match between the two sequences

**Comment by opoudjis** [comment]
This is going to have to be processed as a collection:

* For Preface + stitch(A, B), Preface + A is a valid document; Preface + A + B is not
* Anchors will be shared between A and B, and they will need to be rewritten with prefixes to remain unambiguous --- which collections already do.

What we actually want here is a preprocessing mode in collection processing, which

* identifies the shared initial text in the collection
* injects the appropriate tags and multilingual-rendering attributes in the two language documents, to keep them aligned

**Comment by opoudjis** [comment]
It appears multilingual-rendering to date has been implemented for rendering only for JCGM.

**Comment by opoudjis** [comment]
I am not very enthusiastic about this, but I'm going to implement this in metanorma gem as a postprocessing of Presentation XML. There is code from @Intelligent2013  in JCGM XSLT to handle these tags, but I need to generalise this to HTML and DOC anyway. I will be following his code to arrange elements, including his use of the cross-align element. In fact, I'm going to try and use his XSLT in preprocessing.

**Comment by opoudjis** [comment]
In collections processing, we really need to do without generating PDF of the individual documents; they will not be reused, and are just dead time for document compilation.

**Comment by opoudjis** [comment]
The JCGM XSLT has a model of iterating through the first document in the collection, as a master, and all other documents in the collection, as (slaves) ahem, dependents. We cannot do that, because the first document is likeliest to be a preface: we will need markup in the manifest on the status of each document with bilingual alignment. 

Elements to be aligned are rendered inside `<cross-align/>`, which is populated as an XSL:FO table in JCGM. We will retain that, and process cross-align in HTML (and DOC?)

**Comment by ronaldtse** [comment]
@opoudjis the only true JCGM bilingual document is JCGM 200:
* https://github.com/metanorma/mn-samples-jcgm/issues/24

JCGM 100 is in both English and French but they are published separately.

**Comment by opoudjis** [comment]
I am making up a bilingual out of JCGM 100 at the moment, to see how far I can get in reusing Alex's XSLT, and I will be tinkering with that document.

The JCGM 200 document alternating between one and two columns is irritating, but sadly realistic.

**Comment by opoudjis** [comment]
Performance is better but not great: 47 sec for JCGM 100 (12 MB) to move text in place in parallel columns. Parking code in PR, not yet rendered.

Both I and @Intelligent2013 will need to render Presentation XML `cross-align/align-cell` into parallel table cells. I will need to do so in a single HTML file (since that is what parallel columns ends up requiring.)

**Comment by opoudjis** [comment]
For `<cross-align>` rendering to work in HTML, we are finally going to have to bite the bullet and parse whatever is in Presentation XML in sequence, rather than by query. cross-align takes priority over clauses: it contains them.

**Comment by opoudjis** [comment]
The proof of concept collection is JGCM 100 EN + FR. Am getting bi-column output, but it needs a lot of care, and parsing now needs to be a lot more dumb in just spitting out what it receives in Presentation XML, rather than being opinionated.

**Comment by opoudjis** [comment]
This issue has been on hold too long. I will update the PRs to reflect current code, and merge them.

**Comment by opoudjis** [comment]
Got metanorma working, but it's based on wrapping clauses inside /cross-align/align-cell. Since we uniformly expect clauses to be children of preface and sections, and we render based on displayorder attributes of clauses, it is more sensible to nest cross-align within clauses.

I won't do that for now, but I will copy displayorder to cross-align from the clause it contains, as it is otherwise preventing rendering at all. But nesting clauses within align-cell is counterproductive, I would rather the alignment be based on attributes, as it is in preliminary processing.

The code is currently specific to JCGM, and I think it will need to be rethought. But walking away from this, now that the PRs are being resolved.

**Comment by opoudjis** [comment]
cross-align/align-cell is resulting in titles being stranded away from their clause parents; that's why the HTML rendering does not understand any titles at the moment, because isodoc expects to find titles only as children of clauses:

````
<clause>
<cross-align>
<align-cell>
<title>
````

Will make processing titles more flexible.
